pygfx/objects/_ruler.py

```diff
@@ -0,0 +1,489 @@
+from math import floor, ceil, log10
+
+import numpy as np
+
+from ._base import WorldObject
+from ._more import Line, Points, Text
+from ..resources import Buffer
+from ..geometries import Geometry, TextGeometry
+from ..materials import LineMaterial, PointsMaterial, TextMaterial
+from ..utils.compgeo import get_visible_part_of_line_ndc
+
+
+class Ruler(WorldObject):
+    """An object to represent a ruler with tickmarks.
+
+    Can be used to measure distances in a scene, or as an axis in a plot.
+
+    Usage:
+
+    * Use the properties (most notably ``start_pos`` and ``end_pos``).
+    * Call ``update()`` on each draw.
+    """
+
+    def __init__(
+        self,
+        *,
+        tick_format="0.4g",
+        tick_side="left",
+        min_tick_distance=50,
+        ticks_at_end_points=False,
+    ):
+        super().__init__()
+
+        self.start_pos = 0, 0, 0
+        self.end_pos = 0, 0, 0
+        self.start_value = 0.0
+        self.ticks = None
+
+        self.tick_format = tick_format
+        self.tick_side = tick_side
+        self.min_tick_distance = min_tick_distance
+        self.ticks_at_end_points = ticks_at_end_points
+
+        # Create a line and poins object, with a shared geometry
+        geometry = Geometry(
+            positions=np.zeros((6, 3), np.float32),
+            sizes=np.zeros((6,), np.float32),
+        )
+        self._line = Line(geometry, LineMaterial(color="w", thickness=2))
+        self._points = Points(geometry, PointsMaterial(color="w", size_mode="vertex"))
+        self.add(self._line, self._points)
+        self._text_object_pool = []
+
+        # todo: a material to draw proper tick marks
+        # todo: Text object that draws each line at a position from geometry
+
+    # -- Properties to easily access sub-objects
+
+    @property
+    def line(self):
+        """The line object that shows the ruler's path."""
+        return self._line
+
+    @property
+    def points(self):
+        """The points object that shows the ruler's tickmarks."""
+        return self._points
+
+    # Note: text should also be here eventually
+
+    # -- Main properties
+
+    @property
+    def start_pos(self):
+        """The start posision of the ruler as a 3-tuple or 3-element array.
```

Comment:
in world space I'm assuming?

Reply:
Model space, actually. Will update the docs to make this explicit.